liquidctl/driver/usb.py

```diff
@@ -422,7 +425,40 @@ def write(self, data):
         if not report_id:
             sent += 1
         return sent
-
+        
+    def get_feature_report(self, report_id, length):
+        """Get a feature report from a HID device
+        
+        Upon return, the first byte will contain the Report ID, and
+        the report data itself will begin at the second byte (data[1])
+        """
+        data={}
+        res = self.ctrl_transfer(
+                    bmRequestType=CTRL_TYPE_CLASS | CTRL_RECIPIENT_INTERFACE | ENDPOINT_IN,
+                    bRequest=_HID_GET_REPORT,
+                    wValue=(_HID_FEATURE << 8) | report_id,
+                    wIndex=self.bInterfaceNumber,
+                    data_or_wLength=data)
+        return data
```

Comment:
Okay, let me check the underlying function.

Reply:
I believe the implementation is okay as-is. Even if device does not support numbered reports, the user must specify a Report_ID as an input parameter. This value can be 0. We can see that the underlying implementation always sets `cbuff[0]` to `report_num`  regardless of whether it is zero or non-zero. The client (or caller) of the function must recognize that the first byte in the reply will be the same `report_num` that they specified in the invocation of the function.
```python
  def get_feature_report(self, int report_num, int max_length):
      if self._c_hid == NULL:
          raise ValueError('not open')
      cdef hid_device * c_hid = self._c_hid
      cdef unsigned char lbuff[16]
      cdef unsigned char* cbuff
      cdef size_t c_max_length = max_length
      cdef int n
      if max_length <= 16:
          cbuff = lbuff
      else:
          cbuff = <unsigned char *>malloc(max_length)
      cbuff[0] = report_num
      with nogil:
        n = hid_get_feature_report(c_hid, cbuff, c_max_length)
      res = []
      if n < 0:
          raise IOError('read error')
      for i in range(n):
          res.append(cbuff[i])
      if max_length > 16:
          free(cbuff)
      return res
```

Reply:
The thing is that in that code `hid_get_feature_report` has already normalized the cases with and without a report ID, but we're not yet doing that here.

Take a look at how hidapi implements `hid_get_feature_report` on top of libusb:

```c
int HID_API_EXPORT hid_get_feature_report(hid_device *dev, unsigned char *data, size_t length)
{
	int res = -1;
	int skipped_report_id = 0;
	int report_number = data[0];

	if (report_number == 0x0) {
		/* Offset the return buffer by 1, so that the report ID
		   will remain in byte 0. */
		data++;
		length--;
		skipped_report_id = 1;
	}
	res = libusb_control_transfer(dev->device_handle,
		LIBUSB_REQUEST_TYPE_CLASS|LIBUSB_RECIPIENT_INTERFACE|LIBUSB_ENDPOINT_IN,
		0x01/*HID get_report*/,
		(3/*HID feature*/ << 8) | report_number,
		dev->interface,
		(unsigned char *)data, length,
		1000/*timeout millis*/);

	if (res < 0)
		return -1;

	if (skipped_report_id)
		res++;

	return res;
}
```

The relevant bit is when the `data` pointer is incremented for `report_number == 0x0`, making the response from the control transfer end up from byte index 1 onward.

Reply:
It seems the crucial difference is that `hid_get_feature_report` requires the user to pass in a data buffer (`unsigned char *data`) where the first byte of that buffer is the Report ID.

But in `usb.py` we're passing `report_num` as an integer and we're not passing in a data buffer. Instead, we let the underlying function build the data buffer, which we return to the caller.

So I suppose the question is:

- If the input parameter `report_num` is 0, what will be in the first byte of the returned buffer? Will the first byte be the start of data? 
- If the input parameter `report_num` is > 0, what will be in the first byte of the returned buffer? Will the first byte be the report number? I can answer this one: the first byte is the report number because our USB captures are showing that.

Reply:
Maybe the comment of the function can be changed to this?
```python
    def get_feature_report(self, report_id, length):
        """Get a feature report from a HID device
        
        If report_id > 0, the first byte of the returned data will
        contain the Report ID, and the report data itself will begin at
        the second byte (data[1])
        
        If report_id = 0, the data will begin at the first byte (data[0])
        """
```

Reply:
You're right that the upstream API is designed around C's limitations...

But the most important aspect here is that this API matches whatever we exposed as `HidapiDevice.get_feature_report`, because at runtime time a driver can be using either backend, depending on the platform and the value of the `--hid` option.  And at the moment `HidapiDevice.get_feature_report` has been made to follow the upstream (=hidapi) semantics: response[0] is _always_ the report number.

---

We could discuss changing the semantics of our `get_feature_report` API (in both `PyUsbHid` and `HidapiDevice`), but the only sensible improvement would be to not have the report number in the response at all: the caller already knows which report it is.  Implementing that would still require manipulating the data received during the control transfer, since at the USB level the report number is part of payload _if and only if_ it isn't zero.  So we may choose to (always) omit the report number in our "get report" API, but that would involve _conditionally_ (if `report_id != 0`) shifting out the first byte of the response.

Reply:
PyUsbHid has been removed in the main branch (fbc078bde2f0429eec6ac8cdd24343e579595e52).